contracts/EscrowAndStaking/CrossChainCollector.sol

```diff
@@ -0,0 +1,408 @@
+        uint _revShare
+    ) external {
+        require(msg.sender == stakingThales, "InvSender");
+        if (masterCollector == address(this)) {
+            _storeRewards(
+                masterCollectorChain,
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            ++numOfCollectedResultsForThisPeriod;
+            if (numOfCollectedResultsForThisPeriod == numOfActiveCollectors) {
+                readyToBroadcast = true;
+            }
+        } else {
+            bytes memory message = abi.encode(
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            _sendMessageToChain(masterCollectorChain, message);
+        }
+        emit SentOnClosePeriod(_totalStakedLastPeriodEnd, _totalEscrowedLastPeriodEnd, _bonusPoints, _revShare);
+    }
+
+    function broadcastMessageToAll() external onlyOwner {
+        if (readyToBroadcast) {
+            _broadcastMessageToAll();
+            numOfCollectedResultsForThisPeriod = 0;
+            readyToBroadcast = false;
+            ++period;
+            if (weeklyRewardsDecreaseFactor > 0) {
+                _setRewardsForNextPeriod();
+            }
+        }
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    /// handle a received message
+    function _ccipReceive(Client.Any2EVMMessage memory any2EvmMessage) internal override {
+        // decoding the message
+        if (!messageIdAlreadyReceived[any2EvmMessage.messageId]) {
+            address sender = abi.decode(any2EvmMessage.sender, (address));
+            messagesReceivedFromChainSelector[numOfMessagesReceived] = any2EvmMessage.sourceChainSelector;
+            messagesReceived[numOfMessagesReceived] = any2EvmMessage.data;
+            messageIdAlreadyReceived[any2EvmMessage.messageId] = true;
+            ++numOfMessagesReceived;
+
+            if (masterCollector == address(this)) {
+                uint sourceChainSelector = any2EvmMessage.sourceChainSelector;
+                if (readOnlyMode) {
+                    if (collectorAddress[sourceChainSelector] == sender) {
+                        _calculateRewards(any2EvmMessage.data, sourceChainSelector);
+                    }
+                    ++numOfCollectedResultsForThisPeriod;
+                    if (numOfCollectedResultsForThisPeriod == numOfActiveCollectors) {
+                        readyToBroadcast = true;
+                    }
+                } else if (
+                    collectorAddress[sourceChainSelector] == sender && lastPeriodForChain[sourceChainSelector] < period
+                ) {
+                    lastPeriodForChain[sourceChainSelector] = period;
+                    _calculateRewards(any2EvmMessage.data, sourceChainSelector);
+                    ++numOfCollectedResultsForThisPeriod;
```

Comment:
I am a bit confused as to why this is incremented both on ccipReveive and in the sendMessage method.

Reply:
this should only be maintained on master chain, no?

Reply:
> I am a bit confused as to why this is incremented both on ccipReveive and in the sendMessage method.

because the local Staking contract of the master collector does not enter through the `_ccipReceive()` only through the `sendOnClosePeriod()`

Reply:
> this should only be maintained on master chain, no?

correct, it's maintained only on the master chain